pkg/dataobj/metastore/metastore.go

```diff
@@ -0,0 +1,175 @@
+package metastore
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/backoff"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/thanos-io/objstore"
+
+	"github.com/grafana/loki/v3/pkg/dataobj"
+	"github.com/grafana/loki/v3/pkg/logproto"
+)
+
+const (
+	metastoreWindowSize = 12 * time.Hour
+)
+
+var (
+	// Define our own builder config because metastore objects are significantly smaller.
+	metastoreBuilderCfg = dataobj.BuilderConfig{
+		SHAPrefixSize:     2,
+		TargetObjectSize:  32 * 1024 * 1024,
+		TargetPageSize:    4 * 1024 * 1024,
+		BufferSize:        32 * 1024 * 1024, // 8x page size
+		TargetSectionSize: 4 * 1024 * 1024,  // object size / 8
+	}
+)
+
+type Manager struct {
+	metastoreBuilder *dataobj.Builder
+	tenantID         string
+	metrics          *metastoreMetrics
+	bucket           objstore.Bucket
+	logger           log.Logger
+	backoff          *backoff.Backoff
+
+	builderOnce sync.Once
+}
+
+func NewMetastoreManager(bucket objstore.Bucket, tenantID string, logger log.Logger, reg prometheus.Registerer) (*Manager, error) {
```

Comment:
nit: this can be shortened to `NewManager` without losing any meaning since the caller will be calling `metastore.NewManager`